File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/RadioButton.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.*;
import com.spvessel.spacevil.Decorations.Indents;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.ItemAlignment;
import com.spvessel.spacevil.Flags.KeyCode;

import java.awt.*;
import java.util.List;

/**
 * RadioButton is the basic implementation of a user interface radio button with
 * the ability to be checked (others radio button becomes unchecked) or be
 * unchecked. Contains text and indicator.
 * <p>
 * Supports all events except drag and drop.
 */
public class RadioButton extends Prototype {
    class CustomIndicator extends Indicator {
        @Override
        protected boolean getHoverVerification(float xpos, float ypos) {
            return false;
        }
    }

    private static int count = 0;
    private TextLine _textObject;
    private CustomIndicator _indicator;

    /**
     * Getting indicator item of the .
     * 
     * @return Indicator as com.spvessel.spacevil.Indicator.
     */
    public Indicator getIndicator() {
        return _indicator;
    }

    /**
     * Default  constructor. Text is empty.
     */
    public RadioButton() {
        setItemName("RadioButton_" + count);
        count++;
        eventKeyPress.add(this::onKeyPress);

        // text
        _textObject = new TextLine();
        _textObject.setItemName(getItemName() + "_textObject");

        // indicator
        _indicator = new CustomIndicator();
        _indicator.isFocusable = false;

        setStyle(DefaultsService.getDefaultStyle(RadioButton.class));
    }

    /**
     * Constructs a  with the specified text.
     * 
     * @param text  text as java.lang.String.
     */
    public RadioButton(String text) {
        this();
        setText(text);
    }

    private void onKeyPress(IItem sender, KeyArgs args) {
        if (eventMouseClick != null && (args.key == KeyCode.Enter || args.key == KeyCode.Space)) {
            eventMouseClick.execute(this, new MouseArgs());
        }
    }

    /**
     * Overrided Prototype.setMouseHover(bool) method.
     * <p>
     * Setting this item hovered (mouse cursor located within item's shape).
     * 
     * @param value True: if you want this item be hovered. False: if you want this
     *              item be not hovered.
     */
    @Override
    public void setMouseHover(boolean value) {
        super.setMouseHover(value);
        _indicator.getIndicatorMarker().setMouseHover(value);
        updateState();
    }

    /**
     * Setting alignment of a  text. Combines with alignment by vertically
     * (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(ItemAlignment... alignment) {
        _textObject.setTextAlignment(alignment);
    }

    /**
     * Setting alignment of a  text. Combines with alignment by vertically
     * (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as List of
     *                  com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(List<ItemAlignment> alignment) {
        _textObject.setTextAlignment(alignment);
    }

    /**
     * Setting indents for the text to offset text relative to .
     * 
     * @param margin Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public void setTextMargin(Indents margin) {
        _textObject.setMargin(margin);
    }

    /**
     * Setting indents for the text to offset text relative to .
     * 
     * @param left   Indent on the left.
     * @param top    Indent on the top.
     * @param right  Indent on the right.
     * @param bottom Indent on the bottom.
     */
    public void setTextMargin(int left, int top, int right, int bottom) {
        _textObject.setMargin(left, top, right, bottom);
    }

    /**
     * Getting indents of the text.
     * 
     * @return Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public Indents getTextMargin() {
        return _textObject.getMargin();
    }

    /**
     * Setting font of the text.
     * 
     * @param font Font as java.awt.Font.
     */
    public void setFont(Font font) {
        _textObject.setFont(font);
    }

    /**
     * Setting font size of the text.
     * 
     * @param size New size of the font.
     */
    public void setFontSize(int size) {
        _textObject.setFontSize(size);
    }

    /**
     * Setting font style of the text.
     * 
     * @param style New font style (from java.awt.Font package).
     */
    public void setFontStyle(int style) {
        _textObject.setFontStyle(style);
    }

    /**
     * Setting new font family of the text.
     * 
     * @param fontFamily New font family name.
     */
    public void setFontFamily(String fontFamily) {
        _textObject.setFontFamily(fontFamily);
    }

    /**
     * Getting the current font of the text.
     * 
     * @return Font as java.awt.Font.
     */
    public Font getFont() {
        return _textObject.getFont();
    }

    /**
     * Setting the text.
     * 
     * @param text Text as java.lang.String.
     */
    public void setText(String text) {
        _textObject.setItemText(text);
    }

    /**
     * Setting the text.
     * 
     * @param text Text as java.lang.Object.
     */
    public void setText(Object text) {
        setText(text.toString());
    }

    /**
     * Getting the current text of the .
     * 
     * @return Text as java.lang.String.
     */
    public String getText() {
        return _textObject.getItemText();
    }

    /**
     * Getting the text width (useful when you need resize  by text width).
     * 
     * @return Text width.
     */
    public int getTextWidth() {
        return _textObject.getWidth();
    }

    /**
     * Getting the text height (useful when you need resize  by text
     * height).
     * 
     * @return Text height.
     */
    public int getTextHeight() {
        return _textObject.getHeight();
    }

    /**
     * Setting text color of a .
     * 
     * @param color Text color as java.awt.Color.
     */
    public void setForeground(Color color) {
        _textObject.setForeground(color);
    }

    /**
     * Setting text color of a  in byte RGB format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b) {
        _textObject.setForeground(r, g, b);
    }

    /**
     * Setting background color of an item in byte RGBA format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     * @param a Alpha color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b, int a) {
        _textObject.setForeground(r, g, b, a);
    }

    /**
     * Setting text color of a  in float RGB format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b) {
        _textObject.setForeground(r, g, b);
    }

    /**
     * Setting text color of a  in float RGBA format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     * @param a Alpha color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b, float a) {
        _textObject.setForeground(r, g, b, a);
    }

    /**
     * Getting current text color.
     * 
     * @return Text color as as java.awt.Color.
     */
    public Color getForeground() {
        return _textObject.getForeground();
    }

    /**
     * Initializing all elements in the .
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        // events
        _indicator.getIndicatorMarker().eventToggle = null;
        IMouseMethodState btn_click = (sender, args) -> {
            if (_indicator.getIndicatorMarker().isToggled())
                return;
            _indicator.getIndicatorMarker().setToggled(!_indicator.getIndicatorMarker().isToggled());
            if (_indicator.getIndicatorMarker().isToggled())
                uncheckOthers(sender);
        };
        eventMouseClick.add(btn_click);

        // adding
        addItems(_indicator, _textObject);
    }

    /**
     * Returns True if  is checked otherwise returns False.
     * 
     * @return True:  is checked. False:  is unchecked.
     */
    public boolean isChecked() {
        return _indicator.getIndicatorMarker().isToggled();
    }

    public void setChecked(boolean value) {
        _indicator.getIndicatorMarker().setToggled(value);
    }
/**
     * Setting  checked or unchecked.
     * 
     * @param value True: if you want  to be checked. False: if you want
     *               to be unchecked.
     */
    private void uncheckOthers(IItem sender) {
        List<IBaseItem> items = getParent().getItems();
        for (IBaseItem item : items) {
            if (item instanceof RadioButton && !item.equals(this)) {
                ((RadioButton) item).getIndicator().getIndicatorMarker().setToggled(false);
            }
        }
    }

    /**
     * Setting style of the .
     * <p>
     * Inner styles: "indicator", "text".
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        setForeground(style.foreground);
        setFont(style.font);

        Style innerStyle = style.getInnerStyle("indicator");
        if (innerStyle != null) {
            _indicator.setStyle(innerStyle);
        }
        innerStyle = style.getInnerStyle("text");
        if (innerStyle != null) {
            _textObject.setStyle(innerStyle);
        }
    }
}

File: SpaceVIL/Items/RadioButton.cs
﻿using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// RadioButton is the basic implementation of a user interface radio button with 
    /// the ability to be checked (others radio button becomes unchecked) 
    /// or be unchecked. Contains text and indicator. 
    /// <para/> Supports all events except drag and drop.
    /// </summary>
    public class RadioButton : Prototype
    {
        internal class CustomIndicator : Indicator
        {
            protected internal override bool GetHoverVerification(float xpos, float ypos)
            {
                return false;
            }
        }

        static int count = 0;
        private TextLine _textObject;
        private CustomIndicator _indicator;

        /// <summary>
        /// Getting indicator item of the RadioButton.
        /// </summary>
        /// <returns>Indicator as SpaceVIL.Indicator.</returns>
        public Indicator GetIndicator()
        {
            return _indicator;
        }

        /// <summary>
        /// Default RadioButton constructor. Text is empty.
        /// </summary>
        public RadioButton()
        {
            SetItemName("RadioButton_" + count);
            SetBackground(255, 255, 255, 20);
            SetSpacing(5, 0);
            EventKeyPress += OnKeyPress;
            count++;

            //text
            _textObject = new TextLine();
            // _text_object.IsFocusable = false;
            _textObject.SetItemName(GetItemName() + "_textObject");

            //indicator
            _indicator = new CustomIndicator();
            _indicator.IsFocusable = false;

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.RadioButton)));
        }
        
        /// <summary>
        /// Constructs a RadioButton with the specified text.
        /// </summary>
        /// <param name="text">RadioButton text as System.String.</param>
        public RadioButton(String text) : this()
        {
            SetText(text);
        }

        private void OnKeyPress(object sender, KeyArgs args)
        {
            if (args.Key == KeyCode.Enter || args.Key == KeyCode.Space)
                EventMouseClick?.Invoke(this, new MouseArgs());
        }

        /// <summary>
        /// Overrided Prototype.SetMouseHover(bool) method.
        /// <para/> Setting this item hovered (mouse cursor located within item's shape).
        /// </summary>
        /// <param name="value">True: if you want this item be hovered. 
        /// False: if you want this item be not hovered.</param>
        public override void SetMouseHover(bool value)
        {
            base.SetMouseHover(value);
            _indicator.GetIndicatorMarker().SetMouseHover(IsMouseHover());
            UpdateState();
        }

        /// <summary>
        /// Initializing indicator and text in the RadioButton. 
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            // events
            _indicator.GetIndicatorMarker().EventToggle = null;
            EventMouseClick += (sender, args) => {
                if (_indicator.GetIndicatorMarker().IsToggled())
                {
                    return;
                }
                SetChecked(!IsChecked());
            };

            // adding
            AddItems(_indicator, _textObject);
        }

        /// <summary>
        /// Returns True if RadioButton is checked otherwise returns False.
        /// </summary>
        /// <returns>True: RadioButton is checked. False: RadioButton is unchecked.</returns>
        public bool IsChecked()
        {
            return _indicator.GetIndicatorMarker().IsToggled();
        }

        /// <summary>
        /// Setting RadioButton checked or unchecked.
        /// </summary>
        /// <param name="value">True: if you want RadioButton to be checked. 
        /// False: if you want RadioButton to be unchecked.</param>
        public void SetChecked(bool value)
        {
            var isToggled = IsChecked();
            _indicator.GetIndicatorMarker().SetToggled(value);
            if (value && !isToggled)
            {
                UncheckOthers();
            }
        }

        private void UncheckOthers()
        {
            List<IBaseItem> items = GetParent().GetItems();
            foreach (var item in items)
            {
                if (item is RadioButton && !item.Equals(this))
                {
                    (item as RadioButton).GetIndicator().GetIndicatorMarker().SetToggled(false);
                }
            }
        }
        /// <summary>
        /// Setting alignment of the text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(ItemAlignment alignment)
        {
            _textObject.SetTextAlignment(alignment);
        }
        /// <summary>
        /// Setting alignment of the text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as sequence of SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(params ItemAlignment[] alignment)
        {
            _textObject.SetTextAlignment(alignment);
        }
        /// <summary>
        /// Setting indents for the text to offset text relative to this RadioButton.
        /// </summary>
        /// <param name="margin">Indents as SpaceVIL.Decorations.Indents.</param>
        public void SetTextMargin(Indents margin)
        {
            _textObject.SetMargin(margin);
        }
        /// <summary>
        /// Setting indents for the text to offset text relative to RadioButton.
        /// </summary>
        /// <param name="left">Indent on the left.</param>
        /// <param name="top">Indent on the top.</param>
        /// <param name="right">Indent on the right.</param>
        /// <param name="bottom">Indent on the bottom.</param>
        public void SetTextMargin(int left = 0, int top = 0, int right = 0, int bottom = 0)
        {
            _textObject.SetMargin(left, top, right, bottom);
        }
        /// <summary>
        /// Getting indents of the text.
        /// </summary>
        /// <returns>Indents as SpaceVIL.Decorations.Indents.</returns>
        public Indents GetTextMargin()
        {
            return _textObject.GetMargin();
        }
        /// <summary>
        /// Setting font of the text.
        /// </summary>
        /// <param name="font">Font as System.Drawing.Font.</param>
        public void SetFont(Font font)
        {
            _textObject.SetFont(font);
        }
        /// <summary>
        /// Setting font size of the text.
        /// </summary>
        /// <param name="size">New size of the font.</param>
        public void SetFontSize(int size)
        {
            _textObject.SetFontSize(size);
        }
        /// <summary>
        /// Setting font style of the text.
        /// </summary>
        /// <param name="style">New font style as System.Drawing.FontStyle.</param>
        public void SetFontStyle(FontStyle style)
        {
            _textObject.SetFontStyle(style);
        }
        /// <summary>
        /// Setting new font family of the text.
        /// </summary>
        /// <param name="fontFamily">New font family as System.Drawing.FontFamily.</param>
        public void SetFontFamily(FontFamily fontFamily)
        {
            _textObject.SetFontFamily(fontFamily);
        }
        /// <summary>
        /// Getting the current font of the text.
        /// </summary>
        /// <returns>Font as System.Drawing.Font.</returns>
        public Font GetFont()
        {
            return _textObject.GetFont();
        }

        /// <summary>
        /// Setting the text.
        /// </summary>
        /// <param name="text">Text as System.String.</param>
        public virtual void SetText(String text)
        {
            _textObject.SetItemText(text);
        }

        /// <summary>
        /// Setting the text.
        /// </summary>
        /// <param name="text">Text as System.Object.</param>
        public virtual void SetText(object text)
        {
            SetText(text.ToString());
        }

        /// <summary>
        /// Getting the current text of the RadioButton.
        /// </summary>
        /// <returns>Text as System.String.</returns>
        public virtual String GetText()
        {
            return _textObject.GetItemText();
        }
        /// <summary>
        /// Getting the text width (useful when you need resize RadioButton by text width).
        /// </summary>
        /// <returns>Text width.</returns>
        public int GetTextWidth()
        {
            return _textObject.GetWidth();
        }
        /// <summary>
        /// Getting the text height (useful when you need resize RadioButton by text height).
        /// </summary>
        /// <returns>Text height.</returns>
        public int GetTextHeight()
        {
            return _textObject.GetHeight();
        }
        /// <summary>
        /// Setting text color of a RadioButton.
        /// </summary>
        /// <param name="color">Text color as System.Drawing.Color.</param>
        public void SetForeground(Color color)
        {
            _textObject.SetForeground(color);
        }
        /// <summary>
        /// Setting text color of a RadioButton in byte RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b)
        {
            _textObject.SetForeground(r, g, b);
        }
        /// <summary>
        /// Setting text color of a RadioButton in byte RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        /// <param name="a">Alpha color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b, int a)
        {
            _textObject.SetForeground(r, g, b, a);
        }
        /// <summary>
        /// Setting text color of a RadioButton in float RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b)
        {
            _textObject.SetForeground(r, g, b);
        }
        /// <summary>
        /// Setting text color of a RadioButton in float RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        /// <param name="a">Alpha color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b, float a)
        {
            _textObject.SetForeground(r, g, b, a);
        }
        /// <summary>
        /// Getting current text color.
        /// </summary>
        /// <returns>Text color as System.Drawing.Color.</returns>
        public Color GetForeground()
        {
            return _textObject.GetForeground();
        }

        /// <summary>
        /// Setting style of the RadioButton.
        /// <para/> Inner styles: "indicator", "text".
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);
            Style innerStyle = style.GetInnerStyle("indicator");
            if (innerStyle != null)
            {
                _indicator.SetStyle(innerStyle);
            }
            innerStyle = style.GetInnerStyle("text");
            if (innerStyle != null)
            {
                _textObject.SetStyle(innerStyle);
            }
            SetForeground(style.Foreground);
            SetFont(style.Font);
        }
    }
}
